includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,356 @@
  * @param AbstractObject $object An Islandora Object
  * @param String $filePath video file path
  */
+
+
+libraries_load('aws'); 
+
+use Aws\S3\S3Client;
+use Aws\Exception\AwsException;
+use Aws\S3\ObjectUploader;
+use Aws\S3\MultipartUploader;
+use Aws\Exception\MultipartUploadException;
+
+# Filename of manifest XML file. Any filename is acceptable.
+const MANIFEST_FILE_NAME = '/tmp/upload_manifest_generated.xml';
+
 function islandora_video_streaming_panopto_upload($object, $filePath) {
-  if ($force || !isset[$object['KALTURA_STREAM']) {
+  if ($filePath || !isset($object['PANOPTO_INFO'])) {
+    module_load_include('inc', 'islandora_video', "includes/derivatives" );
     $archival_file = islandora_video_copy_obj_datastream($object);
     if ($archival_file === FALSE) {
       return islandora_video_no_obj_datastream($object->id);
     }
     $filePath = drupal_realpath($archival_file['file']->uri);
     $failureMessage = array(
       'success' => false,
-      'messages' => array(
-        'message' => t('Failed to upload object @pid to Panopto, but no exception caught.'),
+      'messages' => array(),
+    );
+    $streaming = array();
+    
+    $oauthendpoint = variable_get('islandora_video_streaming_panopto_host');
+    $uploadendpoint = $oauthendpoint . '/Panopto/PublicAPI/Rest/sessionUpload';
+    $oauthendpoint .= "/Panopto/oauth2/connect/token";
+    $token = islandora_video_streaming_panopto_get_token($oauthendpoint);
+    $session = null;
+    if ($token) {
+      $session = islandora_video_streaming_panopto_get_session($token, $uploadendpoint); 
+    }
+    else {
+      $messages = array(
+        'message' => t('Failed to retrieve token for object: @pid'),
         'message_sub' => array('@pid' => $object->id),
         'type' => 'watchdog',
         'severity' => WATCHDOG_ERROR,
-      ),
-    );
-
-    //TODO: upload, record result in $streaming
- 
+      );
+      $failureMessage['messages'][] = $messages;
+      return $failureMessage;
+    }
+    $success = false;
+    if ($session) {
+      try {
+        $upload_id = $session['ID'];
+        $upload_target = $session['UploadTarget'];
+        $file_uri = explode("/", $filePath);
+        $FILENAME = end($file_uri);
+        
+        islandora_video_streaming_panopto_create_manifest_for_video($filePath, MANIFEST_FILE_NAME);
+        islandora_video_streaming_panopto_upload_file($upload_target, MANIFEST_FILE_NAME);
```

Comment:
Each of the calls of `islandora_video_streaming_panopto_create_manifest_for_video`, `islandora_video_streaming_panopto_upload_file`, `islandora_video_streaming_panopto_upload_file`, and `islandora_video_streaming_panopto_finish_upload` are clearly dependent on one another, but there is no error checking to enforce this dependency.  This needs to be handled by if conditionals checking return values, or by Exceptions with a `try`...`catch`.

Reply:
Done. I placed` islandora_video_streaming_panopto_upload_file, islandora_video_streaming_panopto_upload_file, and islandora_video_streaming_panopto_finish_upload` inside an `if` statement that checks whether the manifest_file exists. I have also placed the `try and catch` statement within it as `monitor_progress()` relies on the uploading of the files.